crates/router/src/core/locking.rs

```diff
@@ -0,0 +1,191 @@
+use actix_web::rt::time as actix_time;
+use error_stack::{IntoReport, ResultExt};
+use redis_interface as redis;
+use router_env::{instrument, logger, tracing};
+
+use super::errors::{self, RouterResult};
+use crate::routes;
+
+#[derive(Debug)]
+pub enum Lock {
+    Acquired(String),
+    Release(String),
+}
+
+pub struct LockingError {}
+
+#[derive(Clone, Debug)]
+pub enum LockStatus {
+    Acquired(String),
+    AlreadyLocked(String),
+    NotEnabled,
+    AcquireFailedRetriesExhausted(String),
+    Released(String),
+    ReleaseFailedRetriesExhausted(String),
+}
+
+#[derive(Debug)]
+pub enum ActionOnWaitTimeout {
+    Default,
+}
+
+#[derive(Debug)]
+pub struct LockingInput {
+    pub unique_locking_key: String,
+    pub api_identifier: String,
+    pub action_on_wait_timeout: ActionOnWaitTimeout,
+    pub merchant_id: String,
+}
+
+#[instrument(skip(state))]
+pub async fn get_key_and_lock_resource(
+    state: &routes::AppState,
+    locking_input: LockingInput,
+    _should_retry: bool,
+) -> RouterResult<LockStatus> {
+    let api_identifier = locking_input.api_identifier.as_str();
+
+    let is_locking_enabled_for_merchant = true;
+    let is_locking_enabled_on_api = true;
+
+    
+    if is_locking_enabled_for_merchant && is_locking_enabled_on_api {
+        let expiry_in_seconds = 60; // get it from redis
+        let delay_between_retries_in_seconds = 10; // get it from redis
+        let retries = 1; // get from redis based on should_retry, if not present in redis default 1?
+        let locking_key = locking_input.unique_locking_key;
+        lock_resource(
+            state,
+            locking_key,
+            expiry_in_seconds,
+            retries,
+            delay_between_retries_in_seconds,
+            api_identifier,
+        )
+        .await
+    } else {
+        logger::info!(
+            "Resource Locking not Enabled for merchant_id: {} and api: {}",
+            locking_input.merchant_id,
+            api_identifier.to_owned()
+        );
+        Ok(LockStatus::NotEnabled)
+    }
+}
+
+#[instrument(skip(state))]
+pub async fn lock_resource(
+    state: &routes::AppState,
+    locking_key: String,
+    expiry_in_seconds: u64,
+    retries: i32,
+    delay_between_retries_in_seconds: i64,
+    _api_identifier: &str,
+) -> RouterResult<LockStatus> {
+    let redis_key_for_lock = get_redis_key_for_locks(locking_key);
+
+    acquire_lock_on_resource_in_redis(
+        state,
+        redis_key_for_lock.as_str(),
+        true,
+        expiry_in_seconds,
+        delay_between_retries_in_seconds,
+        retries,
+    )
+    .await
+}
+
+#[inline]
+fn get_redis_key_for_locks(a: String) -> String {
```

Comment:
This seems like a unnecessary allocation for input, could either have `&str` or `impl AsRef<str>`

Reply:
Yep. Good catch